Fix `test_object_copy_canned_acl`

This test sets canned ACL to the object. 

```py
def test_object_copy_canned_acl():
    bucket_name = get_new_bucket()
    client = get_client()
    alt_client = get_alt_client()
    client.put_object(Bucket=bucket_name, Key='foo123bar', Body='foo')

    copy_source = {'Bucket': bucket_name, 'Key': 'foo123bar'}
    client.copy_object(Bucket=bucket_name, CopySource=copy_source, Key='bar321foo', ACL='public-read')
    ## check ACL is applied by doing GET from another user
    alt_client.get_object(Bucket=bucket_name, Key='bar321foo')  # <- FAILED HERE

    metadata={'abc': 'def'}
    copy_source = {'Bucket': bucket_name, 'Key': 'bar321foo'}
    client.copy_object(ACL='public-read', Bucket=bucket_name, CopySource=copy_source, Key='foo123bar', Metadata=metadata, MetadataDirective='REPLACE')

    # check ACL is applied by doing GET from another user
    alt_client.get_object(Bucket=bucket_name, Key='foo123bar')
```

ACL is not processed during object copying, therefore alt client fails with 
```
layer/object.go:710     error was transformed   {"request_id": "46f7588e-02a7-47ff-94ff-6d26354983bd", "error": "access denied: access to operation SEARCH is denied by extended ACL check: bearer token is not signed by the container owner"}
handler/util.go:25      could not find object   {"request_id": "46f7588e-02a7-47ff-94ff-6d26354983bd", "method": "GetObject", "bucket_name": "yournamehere-sdcore7ma75qo04d-1", "object_name": "bar321foo", "error": "AccessDenied: 403 => Access Denied."}
```

--- a/api/handler/copy.go
+++ b/api/handler/copy.go
@@ -102,6 +102,7 @@ func (h *handler) CopyObjectHandler(w http.ResponseWriter, r *http.Request) {
 
 	params := &layer.CopyObjectParams{
 		SrcObject:  info,
+		ScrBktInfo: p.BktInfo,
 		DstBktInfo: dstBktInfo,
 		DstObject:  reqInfo.ObjectName,
 		SrcSize:    info.Size,

--- a/api/handler/get.go
+++ b/api/handler/get.go
@@ -168,6 +168,7 @@ func (h *handler) GetObjectHandler(w http.ResponseWriter, r *http.Request) {
 		ObjectInfo: info,
 		Writer:     w,
 		Range:      params,
+		BucketInfo: bktInfo,
 	}
 	if err = h.obj.GetObject(r.Context(), getParams); err != nil {
 		h.logAndSendError(w, "could not get object", reqInfo, err)

--- a/api/handler/get.go
+++ b/api/handler/get.go
@@ -168,7 +168,6 @@ func (h *handler) GetObjectHandler(w http.ResponseWriter, r *http.Request) {
 		ObjectInfo: info,
 		Writer:     w,
 		Range:      params,
-		VersionID:  p.VersionID,
 	}
 	if err = h.obj.GetObject(r.Context(), getParams); err != nil {
 		h.logAndSendError(w, "could not get object", reqInfo, err)

--- a/api/handler/head.go
+++ b/api/handler/head.go
@@ -62,6 +62,7 @@ func (h *handler) HeadObjectHandler(w http.ResponseWriter, r *http.Request) {
 			ObjectInfo: info,
 			Writer:     buffer,
 			Range:      getRangeToDetectContentType(info.Size),
+			BucketInfo: bktInfo,
 		}
 		if err = h.obj.GetObject(r.Context(), getParams); err != nil {
 			h.logAndSendError(w, "could not get object", reqInfo, err, zap.Stringer("oid", info.ID))

--- a/api/handler/head.go
+++ b/api/handler/head.go
@@ -62,7 +62,6 @@ func (h *handler) HeadObjectHandler(w http.ResponseWriter, r *http.Request) {
 			ObjectInfo: info,
 			Writer:     buffer,
 			Range:      getRangeToDetectContentType(info.Size),
-			VersionID:  reqInfo.URL.Query().Get(api.QueryVersionID),
 		}
 		if err = h.obj.GetObject(r.Context(), getParams); err != nil {
 			h.logAndSendError(w, "could not get object", reqInfo, err, zap.Stringer("oid", info.ID))

--- a/api/handler/multipart_upload.go
+++ b/api/handler/multipart_upload.go
@@ -316,6 +316,7 @@ func (h *handler) UploadPartCopy(w http.ResponseWriter, r *http.Request) {
 			Key:      reqInfo.ObjectName,
 		},
 		SrcObjInfo: srcInfo,
+		SrcBktInfo: srcBktInfo,
 		PartNumber: partNumber,
 		Range:      srcRange,
 	}
@@ -380,6 +381,7 @@ func (h *handler) CompleteMultipartUploadHandler(w http.ResponseWriter, r *http.
 		p := &layer.GetObjectParams{
 			ObjectInfo: initPart,
 			Writer:     initPartPayload,
+			BucketInfo: bktInfo,
 		}
 		if err = h.obj.GetObject(r.Context(), p); err != nil {
 			h.logAndSendError(w, "could not get multipart upload acl and/or tagging", reqInfo, err, additional...)

--- a/api/layer/layer.go
+++ b/api/layer/layer.go
@@ -77,6 +77,7 @@ type (
 	GetObjectParams struct {
 		Range      *RangeParams
 		ObjectInfo *data.ObjectInfo
+		BucketInfo *data.BucketInfo
 		Writer     io.Writer
 	}
 
@@ -124,6 +125,7 @@ type (
 	// CopyObjectParams stores object copy request parameters.
 	CopyObjectParams struct {
 		SrcObject  *data.ObjectInfo
+		ScrBktInfo *data.BucketInfo
 		DstBktInfo *data.BucketInfo
 		DstObject  string
 		SrcSize    int64
@@ -382,6 +384,7 @@ func (n *layer) GetObject(ctx context.Context, p *GetObjectParams) error {
 	var params getParams
 
 	params.objInfo = p.ObjectInfo
+	params.bktInfo = p.BucketInfo
 
 	if p.Range != nil {
 		if p.Range.Start > p.Range.End {
@@ -517,6 +520,7 @@ func (n *layer) CopyObject(ctx context.Context, p *CopyObjectParams) (*data.Obje
 			ObjectInfo: p.SrcObject,
 			Writer:     pw,
 			Range:      p.Range,
+			BucketInfo: p.ScrBktInfo,
 		})
 
 		if err = pw.CloseWithError(err); err != nil {

--- a/api/layer/layer.go
+++ b/api/layer/layer.go
@@ -78,7 +78,6 @@ type (
 		Range      *RangeParams
 		ObjectInfo *data.ObjectInfo
 		Writer     io.Writer
-		VersionID  string
 	}
 
 	// HeadObjectParams stores object head request parameters.
@@ -323,10 +322,12 @@ func (n *layer) Owner(ctx context.Context) user.ID {
 	return ownerID
 }
 
-func (n *layer) prepareAuthParameters(ctx context.Context, prm *neofs.PrmAuth) {
+func (n *layer) prepareAuthParameters(ctx context.Context, prm *neofs.PrmAuth, bktOwner user.ID) {
 	if bd, ok := ctx.Value(api.BoxData).(*accessbox.Box); ok && bd != nil && bd.Gate != nil {
-		prm.BearerToken = bd.Gate.BearerToken
-		return
+		if issuer, ok := bd.Gate.BearerToken.Issuer(); ok && bktOwner.Equals(issuer) {
+			prm.BearerToken = bd.Gate.BearerToken
+			return
+		}
 	}
 
 	prm.PrivateKey = &n.anonKey.Key.PrivateKey
@@ -380,8 +381,7 @@ func (n *layer) ListBuckets(ctx context.Context) ([]*data.BucketInfo, error) {
 func (n *layer) GetObject(ctx context.Context, p *GetObjectParams) error {
 	var params getParams
 
-	params.oid = p.ObjectInfo.ID
-	params.cid = p.ObjectInfo.CID
+	params.objInfo = p.ObjectInfo
 
 	if p.Range != nil {
 		if p.Range.Start > p.Range.End {
@@ -584,7 +584,7 @@ func (n *layer) deleteObject(ctx context.Context, bkt *data.BucketInfo, settings
 	}
 
 	for _, id := range ids {
-		if err = n.objectDelete(ctx, bkt.CID, id); err != nil {
+		if err = n.objectDelete(ctx, bkt, id); err != nil {
 			obj.Error = err
 			return obj
 		}

--- a/api/layer/multipart_upload.go
+++ b/api/layer/multipart_upload.go
@@ -208,7 +208,7 @@ func (x *multiObjectReader) Read(p []byte) (n int, err error) {
 		return n, io.EOF
 	}
 
-	x.prm.oid = x.parts[0].ID
+	x.prm.objInfo = x.parts[0]
 
 	x.curReader, err = x.layer.initObjectPayloadReader(x.ctx, x.prm)
 	if err != nil {
@@ -306,8 +306,6 @@ func (n *layer) CompleteMultipartUpload(ctx context.Context, p *CompleteMultipar
 		parts: parts,
 	}
 
-	r.prm.cid = p.Info.Bkt.CID
-
 	obj, err = n.PutObject(ctx, &PutObjectParams{
 		BktInfo: p.Info.Bkt,
 		Object:  p.Info.Key,
@@ -327,7 +325,7 @@ func (n *layer) CompleteMultipartUpload(ctx context.Context, p *CompleteMultipar
 		if partNum == 0 {
 			continue
 		}
-		if err = n.objectDelete(ctx, p.Info.Bkt.CID, objInfo.ID); err != nil {
+		if err = n.objectDelete(ctx, p.Info.Bkt, objInfo.ID); err != nil {
 			n.log.Warn("could not delete upload part",
 				zap.Stringer("object id", objInfo.ID),
 				zap.Stringer("bucket id", p.Info.Bkt.CID),
@@ -347,7 +345,7 @@ func (n *layer) ListMultipartUploads(ctx context.Context, p *ListMultipartUpload
 
 	f := &findParams{
 		attr: [2]string{UploadPartNumberAttributeName, "0"},
-		cid:  p.Bkt.CID,
+		bkt:  p.Bkt,
 	}
 
 	ids, err := n.objectSearch(ctx, f)
@@ -359,7 +357,7 @@ func (n *layer) ListMultipartUploads(ctx context.Context, p *ListMultipartUpload
 	uniqDirs := make(map[string]struct{})
 
 	for i := range ids {
-		meta, err := n.objectHead(ctx, p.Bkt.CID, ids[i])
+		meta, err := n.objectHead(ctx, p.Bkt, ids[i])
 		if err != nil {
 			n.log.Warn("couldn't head object",
 				zap.Stringer("object id", &ids[i]),
@@ -419,7 +417,7 @@ func (n *layer) AbortMultipartUpload(ctx context.Context, p *UploadInfoParams) e
 	}
 
 	for _, info := range objects {
-		err := n.objectDelete(ctx, info.CID, info.ID)
+		err := n.objectDelete(ctx, p.Bkt, info.ID)
 		if err != nil {
 			return err
 		}
@@ -492,7 +490,7 @@ func (n *layer) getUploadParts(ctx context.Context, p *UploadInfoParams) (map[in
 	// and search in attributes by prefix is not supported
 	f := &findParams{
 		attr: [2]string{UploadIDAttributeName, p.UploadID},
-		cid:  p.Bkt.CID,
+		bkt:  p.Bkt,
 	}
 
 	ids, err := n.objectSearch(ctx, f)
@@ -503,7 +501,7 @@ func (n *layer) getUploadParts(ctx context.Context, p *UploadInfoParams) (map[in
 	res := make(map[int]*data.ObjectInfo)
 
 	for i := range ids {
-		meta, err := n.objectHead(ctx, p.Bkt.CID, ids[i])
+		meta, err := n.objectHead(ctx, p.Bkt, ids[i])
 		if err != nil {
 			n.log.Warn("couldn't head a part of upload",
 				zap.Stringer("object id", &ids[i]),

--- a/api/layer/multipart_upload.go
+++ b/api/layer/multipart_upload.go
@@ -51,6 +51,7 @@ type (
 	UploadCopyParams struct {
 		Info       *UploadInfoParams
 		SrcObjInfo *data.ObjectInfo
+		SrcBktInfo *data.BucketInfo
 		PartNumber int
 		Range      *RangeParams
 	}
@@ -166,6 +167,7 @@ func (n *layer) UploadPartCopy(ctx context.Context, p *UploadCopyParams) (*data.
 			ObjectInfo: p.SrcObjInfo,
 			Writer:     pw,
 			Range:      p.Range,
+			BucketInfo: p.SrcBktInfo,
 		})
 
 		if err = pw.CloseWithError(err); err != nil {
@@ -305,6 +307,8 @@ func (n *layer) CompleteMultipartUpload(ctx context.Context, p *CompleteMultipar
 		layer: n,
 		parts: parts,
 	}
+
+	r.prm.bktInfo = p.Info.Bkt
 
 	obj, err = n.PutObject(ctx, &PutObjectParams{
 		BktInfo: p.Info.Bkt,

--- a/api/layer/object.go
+++ b/api/layer/object.go
@@ -26,7 +26,7 @@ import (
 type (
 	findParams struct {
 		attr   [2]string
-		cid    cid.ID
+		bkt    *data.BucketInfo
 		prefix string
 	}
 
@@ -34,8 +34,7 @@ type (
 		// payload range
 		off, ln uint64
 
-		cid cid.ID
-		oid oid.ID
+		objInfo *data.ObjectInfo
 	}
 
 	// ListObjectsParamsCommon contains common parameters for ListObjectsV1 and ListObjectsV2.
@@ -68,10 +67,10 @@ type (
 	}
 )
 
-func (n *layer) objectSearchByName(ctx context.Context, cnr cid.ID, filename string) ([]oid.ID, error) {
+func (n *layer) objectSearchByName(ctx context.Context, bktInfo *data.BucketInfo, filename string) ([]oid.ID, error) {
 	f := &findParams{
 		attr: [2]string{object.AttributeFileName, filename},
-		cid:  cnr,
+		bkt:  bktInfo,
 	}
 	return n.objectSearch(ctx, f)
 }
@@ -79,12 +78,12 @@ func (n *layer) objectSearchByName(ctx context.Context, cnr cid.ID, filename str
 // objectSearch returns all available objects by search params.
 func (n *layer) objectSearch(ctx context.Context, p *findParams) ([]oid.ID, error) {
 	prm := neofs.PrmObjectSelect{
-		Container:      p.cid,
+		Container:      p.bkt.CID,
 		ExactAttribute: p.attr,
 		FilePrefix:     p.prefix,
 	}
 
-	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, p.bkt.Owner)
 
 	res, err := n.neoFS.SelectObjects(ctx, prm)
 
@@ -99,14 +98,14 @@ func newAddress(cnr cid.ID, obj oid.ID) oid.Address {
 }
 
 // objectHead returns all object's headers.
-func (n *layer) objectHead(ctx context.Context, idCnr cid.ID, idObj oid.ID) (*object.Object, error) {
+func (n *layer) objectHead(ctx context.Context, bktInfo *data.BucketInfo, idObj oid.ID) (*object.Object, error) {
 	prm := neofs.PrmObjectRead{
-		Container:  idCnr,
+		Container:  bktInfo.CID,
 		Object:     idObj,
 		WithHeader: true,
 	}
 
-	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, bktInfo.Owner)
 
 	res, err := n.neoFS.ReadObject(ctx, prm)
 	if err != nil {
@@ -120,13 +119,19 @@ func (n *layer) objectHead(ctx context.Context, idCnr cid.ID, idObj oid.ID) (*ob
 // Zero range corresponds to full payload (panics if only offset is set).
 func (n *layer) initObjectPayloadReader(ctx context.Context, p getParams) (io.Reader, error) {
 	prm := neofs.PrmObjectRead{
-		Container:    p.cid,
-		Object:       p.oid,
+		Container:    p.objInfo.CID,
+		Object:       p.objInfo.ID,
 		WithPayload:  true,
 		PayloadRange: [2]uint64{p.off, p.ln},
 	}
 
-	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+	// should be taken from cache
+	bktInfo, err := n.GetBucketInfo(ctx, p.objInfo.Bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, bktInfo.Owner)
 
 	res, err := n.neoFS.ReadObject(ctx, prm)
 	if err != nil {
@@ -137,15 +142,15 @@ func (n *layer) initObjectPayloadReader(ctx context.Context, p getParams) (io.Re
 }
 
 // objectGet returns an object with payload in the object.
-func (n *layer) objectGet(ctx context.Context, addr oid.Address) (*object.Object, error) {
+func (n *layer) objectGet(ctx context.Context, bktInfo *data.BucketInfo, objID oid.ID) (*object.Object, error) {
 	prm := neofs.PrmObjectRead{
-		Container:   addr.Container(),
-		Object:      addr.Object(),
+		Container:   bktInfo.CID,
+		Object:      objID,
 		WithHeader:  true,
 		WithPayload: true,
 	}
 
-	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, bktInfo.Owner)
 
 	res, err := n.neoFS.ReadObject(ctx, prm)
 	if err != nil {
@@ -197,7 +202,7 @@ func (n *layer) PutObject(ctx context.Context, p *PutObjectParams) (*data.Object
 		}
 	}
 
-	id, hash, err := n.objectPutAndHash(ctx, prm)
+	id, hash, err := n.objectPutAndHash(ctx, prm, p.BktInfo)
 	if err != nil {
 		return nil, err
 	}
@@ -228,7 +233,7 @@ func (n *layer) PutObject(ctx context.Context, p *PutObjectParams) (*data.Object
 	n.listsCache.CleanCacheEntriesContainingObject(p.Object, p.BktInfo.CID)
 
 	for _, id := range idsToDeleteArr {
-		if err = n.objectDelete(ctx, p.BktInfo.CID, id); err != nil {
+		if err = n.objectDelete(ctx, p.BktInfo, id); err != nil {
 			n.log.Warn("couldn't delete object",
 				zap.Stringer("version id", id),
 				zap.Error(err))
@@ -355,7 +360,7 @@ func (n *layer) headLastVersionIfNotDeleted(ctx context.Context, bkt *data.Bucke
 }
 
 func (n *layer) headVersions(ctx context.Context, bkt *data.BucketInfo, objectName string) (*objectVersions, error) {
-	ids, err := n.objectSearchByName(ctx, bkt.CID, objectName)
+	ids, err := n.objectSearchByName(ctx, bkt, objectName)
 	if err != nil {
 		return nil, err
 	}
@@ -366,7 +371,7 @@ func (n *layer) headVersions(ctx context.Context, bkt *data.BucketInfo, objectNa
 	}
 
 	for i := range ids {
-		meta, err := n.objectHead(ctx, bkt.CID, ids[i])
+		meta, err := n.objectHead(ctx, bkt, ids[i])
 		if err != nil {
 			n.log.Warn("couldn't head object",
 				zap.Stringer("object id", &ids[i]),
@@ -415,7 +420,7 @@ func (n *layer) headVersion(ctx context.Context, bkt *data.BucketInfo, p *HeadOb
 		return objInfoFromMeta(bkt, headInfo), nil
 	}
 
-	meta, err := n.objectHead(ctx, bkt.CID, id)
+	meta, err := n.objectHead(ctx, bkt, id)
 	if err != nil {
 		if client.IsErrObjectNotFound(err) {
 			return nil, apiErrors.GetAPIError(apiErrors.ErrNoSuchVersion)
@@ -437,23 +442,23 @@ func (n *layer) headVersion(ctx context.Context, bkt *data.BucketInfo, p *HeadOb
 }
 
 // objectDelete puts tombstone object into neofs.
-func (n *layer) objectDelete(ctx context.Context, idCnr cid.ID, idObj oid.ID) error {
+func (n *layer) objectDelete(ctx context.Context, bktInfo *data.BucketInfo, idObj oid.ID) error {
 	prm := neofs.PrmObjectDelete{
-		Container: idCnr,
+		Container: bktInfo.CID,
 		Object:    idObj,
 	}
 
-	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, bktInfo.Owner)
 
-	n.objCache.Delete(newAddress(idCnr, idObj))
+	n.objCache.Delete(newAddress(bktInfo.CID, idObj))
 
 	return n.transformNeofsError(ctx, n.neoFS.DeleteObject(ctx, prm))
 }
 
 // objectPutAndHash prepare auth parameters and invoke neofs.CreateObject.
 // Returns object ID and payload sha256 hash.
-func (n *layer) objectPutAndHash(ctx context.Context, prm neofs.PrmObjectCreate) (*oid.ID, []byte, error) {
-	n.prepareAuthParameters(ctx, &prm.PrmAuth)
+func (n *layer) objectPutAndHash(ctx context.Context, prm neofs.PrmObjectCreate, bktInfo *data.BucketInfo) (*oid.ID, []byte, error) {
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, bktInfo.Owner)
 	hash := sha256.New()
 	prm.Payload = wrapReader(prm.Payload, 64*1024, func(buf []byte) {
 		hash.Write(buf)
@@ -564,7 +569,7 @@ func (n *layer) getAllObjectsVersions(ctx context.Context, bkt *data.BucketInfo,
 	ids := n.listsCache.Get(cacheKey)
 
 	if ids == nil {
-		ids, err = n.objectSearch(ctx, &findParams{cid: bkt.CID, prefix: prefix})
+		ids, err = n.objectSearch(ctx, &findParams{bkt: bkt, prefix: prefix})
 		if err != nil {
 			return nil, err
 		}
@@ -576,7 +581,7 @@ func (n *layer) getAllObjectsVersions(ctx context.Context, bkt *data.BucketInfo,
 	versions := make(map[string]*objectVersions, len(ids)/2)
 
 	for i := 0; i < len(ids); i++ {
-		obj := n.objectFromObjectsCacheOrNeoFS(ctx, bkt.CID, ids[i])
+		obj := n.objectFromObjectsCacheOrNeoFS(ctx, bkt, ids[i])
 		if obj == nil {
 			continue
 		}
@@ -680,13 +685,13 @@ func (n *layer) isVersioningEnabled(ctx context.Context, bktInfo *data.BucketInf
 	return settings.VersioningEnabled
 }
 
-func (n *layer) objectFromObjectsCacheOrNeoFS(ctx context.Context, cnr cid.ID, obj oid.ID) *object.Object {
+func (n *layer) objectFromObjectsCacheOrNeoFS(ctx context.Context, bktInfo *data.BucketInfo, obj oid.ID) *object.Object {
 	var (
 		err  error
-		meta = n.objCache.Get(newAddress(cnr, obj))
+		meta = n.objCache.Get(newAddress(bktInfo.CID, obj))
 	)
 	if meta == nil {
-		meta, err = n.objectHead(ctx, cnr, obj)
+		meta, err = n.objectHead(ctx, bktInfo, obj)
 		if err != nil {
 			n.log.Warn("could not fetch object meta", zap.Error(err))
 			return nil

--- a/api/layer/object.go
+++ b/api/layer/object.go
@@ -35,6 +35,7 @@ type (
 		off, ln uint64
 
 		objInfo *data.ObjectInfo
+		bktInfo *data.BucketInfo
 	}
 
 	// ListObjectsParamsCommon contains common parameters for ListObjectsV1 and ListObjectsV2.
@@ -125,13 +126,7 @@ func (n *layer) initObjectPayloadReader(ctx context.Context, p getParams) (io.Re
 		PayloadRange: [2]uint64{p.off, p.ln},
 	}
 
-	// should be taken from cache
-	bktInfo, err := n.GetBucketInfo(ctx, p.objInfo.Bucket)
-	if err != nil {
-		return nil, err
-	}
-
-	n.prepareAuthParameters(ctx, &prm.PrmAuth, bktInfo.Owner)
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, p.bktInfo.Owner)
 
 	res, err := n.neoFS.ReadObject(ctx, prm)
 	if err != nil {

--- a/api/layer/system_object.go
+++ b/api/layer/system_object.go
@@ -14,7 +14,6 @@ import (
 	"github.com/nspcc-dev/neofs-s3-gw/api/errors"
 	"github.com/nspcc-dev/neofs-s3-gw/api/layer/neofs"
 	"github.com/nspcc-dev/neofs-sdk-go/object"
-	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
 	"go.uber.org/zap"
 )
 
@@ -59,7 +58,7 @@ func (n *layer) HeadSystemObject(ctx context.Context, bkt *data.BucketInfo, objN
 func (n *layer) DeleteSystemObject(ctx context.Context, bktInfo *data.BucketInfo, name string) error {
 	f := &findParams{
 		attr: [2]string{objectSystemAttributeName, name},
-		cid:  bktInfo.CID,
+		bkt:  bktInfo,
 	}
 	ids, err := n.objectSearch(ctx, f)
 	if err != nil {
@@ -68,7 +67,7 @@ func (n *layer) DeleteSystemObject(ctx context.Context, bktInfo *data.BucketInfo
 
 	n.systemCache.Delete(systemObjectKey(bktInfo, name))
 	for i := range ids {
-		if err = n.objectDelete(ctx, bktInfo.CID, ids[i]); err != nil {
+		if err = n.objectDelete(ctx, bktInfo, ids[i]); err != nil {
 			return err
 		}
 	}
@@ -118,7 +117,7 @@ func (n *layer) putSystemObjectIntoNeoFS(ctx context.Context, p *PutSystemObject
 		prm.Attributes = append(prm.Attributes, [2]string{k, v})
 	}
 
-	id, hash, err := n.objectPutAndHash(ctx, prm)
+	id, hash, err := n.objectPutAndHash(ctx, prm, p.BktInfo)
 	if err != nil {
 		return nil, err
 	}
@@ -132,7 +131,7 @@ func (n *layer) putSystemObjectIntoNeoFS(ctx context.Context, p *PutSystemObject
 	}
 
 	for _, id := range idsToDeleteArr {
-		if err = n.objectDelete(ctx, p.BktInfo.CID, id); err != nil {
+		if err = n.objectDelete(ctx, p.BktInfo, id); err != nil {
 			n.log.Warn("couldn't delete system object",
 				zap.Stringer("version id", id),
 				zap.String("name", p.ObjName),
@@ -168,12 +167,7 @@ func (n *layer) getSystemObjectFromNeoFS(ctx context.Context, bkt *data.BucketIn
 
 	objInfo := versions.getLast()
 
-	var addr oid.Address
-
-	addr.SetContainer(bkt.CID)
-	addr.SetObject(objInfo.ID)
-
-	obj, err := n.objectGet(ctx, addr)
+	obj, err := n.objectGet(ctx, bkt, objInfo.ID)
 	if err != nil {
 		return nil, err
 	}
@@ -214,7 +208,7 @@ func (n *layer) getCORS(ctx context.Context, bkt *data.BucketInfo, sysName strin
 func (n *layer) headSystemVersions(ctx context.Context, bkt *data.BucketInfo, sysName string) (*objectVersions, error) {
 	f := &findParams{
 		attr: [2]string{objectSystemAttributeName, sysName},
-		cid:  bkt.CID,
+		bkt:  bkt,
 	}
 	ids, err := n.objectSearch(ctx, f)
 	if err != nil {
@@ -223,7 +217,7 @@ func (n *layer) headSystemVersions(ctx context.Context, bkt *data.BucketInfo, sy
 
 	versions := newObjectVersions(sysName)
 	for i := range ids {
-		meta, err := n.objectHead(ctx, bkt.CID, ids[i])
+		meta, err := n.objectHead(ctx, bkt, ids[i])
 		if err != nil {
 			n.log.Warn("couldn't head object",
 				zap.Stringer("object id", &ids[i]),

--- a/api/layer/versioning_test.go
+++ b/api/layer/versioning_test.go
@@ -50,7 +50,6 @@ func (tc *testContext) getObject(objectName, versionID string, needError bool) (
 	err = tc.layer.GetObject(tc.ctx, &GetObjectParams{
 		ObjectInfo: objInfo,
 		Writer:     content,
-		VersionID:  versionID,
 	})
 	require.NoError(tc.t, err)
 

--- a/api/layer/versioning_test.go
+++ b/api/layer/versioning_test.go
@@ -50,6 +50,7 @@ func (tc *testContext) getObject(objectName, versionID string, needError bool) (
 	err = tc.layer.GetObject(tc.ctx, &GetObjectParams{
 		ObjectInfo: objInfo,
 		Writer:     content,
+		BucketInfo: tc.bktInfo,
 	})
 	require.NoError(tc.t, err)
 

